File: core/src/main/kotlin/org/bitlap/core/sql/MDColumn.kt
/* Copyright (c) 2022 bitlap.org */
package org.bitlap.core.sql

import org.apache.calcite.sql.SqlAggFunction
import org.apache.calcite.sql.`fun`.SqlCountAggFunction
import org.apache.calcite.sql.`fun`.SqlSumAggFunction
import org.bitlap.common.utils.PreConditions

/**
 * Analyse metric & dimension model from sql select node
 */
data class MDColumn(val name: String, val type: ColumnType) {

    var project: Boolean = false
    var filter: Boolean = false

    /**
     * if column is pure with no function call
     */
    var pure: Boolean = false

    /**
     * left: agg function
     * right: function quantifier, for example distinct and etc.
     */
    val aggregates = mutableSetOf<Pair<SqlAggFunction, String>>()

    fun checkType(targetType: ColumnType): MDColumn {
        PreConditions.checkExpression(
            type == targetType,
            msg = "Unable to judge Column $name target type: $targetType should equals to current type: $type"
        )
        return this
    }

    fun withFilter(f: Boolean) = this.also { it.filter = f }
    fun withProject(p: Boolean) = this.also { it.project = p }
    fun withPure(p: Boolean) = this.also { it.pure = p }
    fun withAgg(vararg agg: Pair<SqlAggFunction, String>) = this.also { it.aggregates.addAll(agg) }
    fun withAgg(agg: Set<Pair<SqlAggFunction, String>>) = this.also { it.aggregates.addAll(agg) }

    fun isDistinct(): Boolean {
        if (this.aggregates.isEmpty()) {
            return false
        }
        return this.aggregates.filter { it.first is SqlCountAggFunction }
            .any { it.second.lowercase() == "distinct" }
    }

    fun isSum(): Boolean {
        if (this.aggregates.isEmpty()) {
            return false
        }
        return this.aggregates.any { it.first is SqlSumAggFunction }
    }

    override fun toString(): String {
        return "MDColumn(name='$name', type=$type, project=$project, filter=$filter, pure=$pure, aggregates=$aggregates)"
    }
}

sealed interface ColumnType
object MetricCol : ColumnType
object DimensionCol : ColumnType

File: bitlap-core/src/main/scala/org/bitlap/core/sql/MDColumn.scala
/*
 * Copyright 2020-2023 IceMimosa, jxnu-liguobin and the Bitlap Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.bitlap.core.sql

import scala.collection.mutable

import org.bitlap.common.extension._
import org.bitlap.common.utils.PreConditions

import org.apache.calcite.sql.`fun`.SqlCountAggFunction
import org.apache.calcite.sql.`fun`.SqlSumAggFunction
import org.apache.calcite.sql.SqlAggFunction

/** Analyse metric & dimension model from sql select node
 */
final case class MDColumn(name: String, `type`: ColumnType) {

  var project: Boolean = false
  var filter: Boolean  = false

  /** if column is pure with no function call
   */
  var pure: Boolean = false

  /** left: agg function right: function quantifier, for example distinct and etc.
   */
  val aggregates: mutable.Set[(SqlAggFunction, String)] = scala.collection.mutable.Set[(SqlAggFunction, String)]()

  def checkType(targetType: ColumnType): MDColumn = {
    PreConditions.checkExpression(
      `type` == targetType,
      "",
      s"Unable to judge Column $name target type: $targetType should equals to current type: ${`type`}"
    )
    this
  }

  def withFilter(f: Boolean): this.type                      = this.also { it => it.filter = f }
  def withProject(p: Boolean): this.type                     = this.also { it => it.project = p }
  def withPure(p: Boolean): this.type                        = this.also { it => it.pure = p }
  def withAgg(agg: (SqlAggFunction, String)*): this.type     = this.also { it => it.aggregates.addAll(agg) }
  def withAgg(agg: Set[(SqlAggFunction, String)]): this.type = this.also { it => it.aggregates.addAll(agg) }

  def isDistinct: Boolean = {
    if (this.aggregates.isEmpty) {
      return false
    }
    this.aggregates.filter { it => it._1.isInstanceOf[SqlCountAggFunction] }.exists { it =>
      it._2.toLowerCase() == "distinct"
    }
  }

  def isSum: Boolean = {
    if (this.aggregates.isEmpty) {
      return false
    }
    this.aggregates.exists { it => it._1.isInstanceOf[SqlSumAggFunction] }
  }

  override def toString: String = {
    s"MDColumn(name='$name', type=${`type`}, project=$project, filter=$filter, pure=$pure, aggregates=$aggregates)"
  }
}

sealed trait ColumnType
object MetricCol    extends ColumnType
object DimensionCol extends ColumnType
